coordinator/src/escrow_coordinator/mod.rs

```diff
@@ -29,52 +31,66 @@ impl EscrowCoordinator {
     }
 
     pub async fn run(&mut self) -> anyhow::Result<()> {
+        let my_pubkey = self.nostr_client.public_key();
         let filter_note = Filter::new()
-            .kind(Kind::EncryptedDirectMessage)
-            .custom_tag(
-                SingleLetterTag::lowercase(Alphabet::P),
-                [PublicKey::from_bech32(&self.nostr_client.get_npub()?)?.to_hex()],
-            )
-            .since(Timestamp::now());
+            .kind(Kind::GiftWrap)
+            .pubkey(my_pubkey)
+            .limit(0);
 
         self.nostr_client
             .client
             .subscribe(vec![filter_note], None)
-            .await;
+            .await?;
         let mut notifications = self.nostr_client.client.notifications();
 
-        while let Ok(notification) = notifications.recv().await {
-            if let RelayPoolNotification::Event { event, .. } = notification {
-                if let Some(decrypted) = self
-                    .nostr_client
-                    .decrypt_msg(&event.content, &event.author())
-                    .await
-                {
-                    dbg!("Received event: {:?}", &decrypted);
-                    if let Ok((contract_hash, contract)) = self.parse(decrypted.as_str()).await {
-                        if self.pending_contracts.contains_key(&contract_hash) {
-                            self.pending_contracts.remove(&contract_hash);
-                            self.begin_trade(&contract_hash, &contract).await?;
-                        } else {
-                            self.pending_contracts.insert(contract_hash, contract);
+        loop {
+            match notifications.recv().await {
+                Ok(notification) => {
+                    if let RelayPoolNotification::Event { event, .. } = notification {
```

Comment:
Maybe we should keep some state on previously received events using a HashSet of Event IDs? I noticed sometimes events seem to get recieved twice, if this happens here we could for example interpret the same event twice as a full contract (e.g. interpreting duplicate seller event twice once as seller, once as buyer) which would interrupt the trade. Maybe a malicious relay could offer the same event multiple times as DoS? Alternatively we could also keep state of which npub submitted the contract so we can differentiate between seller and buyer submissions, but just checking event ids seems easier.

Reply:
Addressed in separate Issue https://github.com/f321x/cashu-escrow-kit/issues/12